packaging/setup/plugins/ovirt-engine-common/ovirt-engine-keycloak/core/ovirt_internal.py

```diff
@@ -0,0 +1,490 @@
+
+    @plugin.event(
+        stage=plugin.Stages.STAGE_CLOSEUP,
+        after=(
+            oengcommcons.Stages.CORE_ENGINE_START,
+        ),
+        condition=lambda self: (
+            self.environment[oenginecons.CoreEnv.ENABLE] and
+            self.environment[oenginecons.EngineDBEnv.NEW_DATABASE] and
+            not self.environment[osetupcons.CoreEnv.DEVELOPER_MODE]
+        ),
+    )
+    def _setup_ovirt(self):
+        password = self.environment.get(oenginecons.ConfigEnv.ADMIN_PASSWORD)
+        if password:
+            # TODO sometimes keycloak app is not ready soon enough  resulting
+            # in 503 error on attempt to use kcadm.sh / api
+            # This sleep is only a workaround until better way is found
+            time.sleep(30)
+            self.logger.info('Start with setting up Keycloak for Ovirt Engine')
+            self._setup_keystore()
+
+            self.logger.info(_('Logging in with Keycloak CLI admin'))
+            self._login(
+                passwd=password
+            )
+
+            self.logger.info('Creating realm in with Keycloak CLI admin')
+            self._setup_realm()
+
+            # please note that client_id is an id (hash) and
+            # is not the human readable
+            # 'client id' defined as
+            # okkcons.Const.KEYCLOAK_INTERNAL_CLIENT_NAME
+            self.logger.info(
+                'Creating internal client in with Keycloak CLI admin'
+            )
+            client_id = self._setup_client()
+
+            for scope in okkcons.Const.OVIRT_CLIENT_SCOPES:
+                self._setup_client_scope(
+                    scope_name=scope,
+                    client_id=client_id,
+                )
+
+            self._setup_protocol_mapper_username(
+                client_id=client_id
+            )
+            self._setup_protocol_mapper_groups(
+                client_id=client_id
+            )
+
+            administrator_group_id = self._setup_ovirt_administrator_group()
+            self._setup_internal_admin_user(
+                administrator_group_id=administrator_group_id,
+                password=password,
+            )
+            self.logger.info('Done with setting up Keycloak for Ovirt Engine')
+
+    def _setup_keystore(self):
+        self.execute(
+            (
+                okkcons.FileLocations.KEYCLOAK_CLI_ADMIN_SCRIPT,
+                'config',
+                'truststore',
+                '--trustpass', oenginecons.Const.PKI_PASSWORD,
+                oenginecons.FileLocations.OVIRT_ENGINE_PKI_ENGINE_TRUST_STORE
+            ),
+        )
+
+    def _login(self, passwd):
+        engine_host = self.environment[
+            oenginecons.ConfigEnv.ENGINE_FQDN
+        ]
+
+        keycloak_server_url = os.path.join(
+            engine_host,
+            okkcons.Const.KEYCLOAK_WEB_CONTEXT,
+        )
+
+        self.execute(
+            (
+                okkcons.FileLocations.KEYCLOAK_CLI_ADMIN_SCRIPT,
```

Comment:
ad 1) Yes - I see it can be stored elsewhere.
ad 2) Good catch - I can't see any explicit way to logout, in worst case  I can cleanup config stored at 1)  which should have a bit similar effect. kcadm.sh tool would need to be re-authenticated ... but JWT token obtained earlier will be still valid until it expires.

Reply:
I think we can postpone that as the configuration contains only a token that will eventually expire.  Created an issue #17.